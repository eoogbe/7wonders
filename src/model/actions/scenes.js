/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const VIEW_WONDER_CHOOSER = 'VIEW_WONDER_CHOOSER';

export const VIEW_WONDER_PANEL = 'VIEW_WONDER_PANEL';

export const VIEW_GAME_BOARD = 'VIEW_GAME_BOARD';

export const START_AGE = 'START_AGE';

export const SCORE_GAME = 'SCORE_GAME';

export const QUIT = 'QUIT';

export const viewWonderChooser = (playerId) => ({
  type: VIEW_WONDER_CHOOSER,
  payload: {
    playerId,
  },
});

export const viewWonderPanel = () => ({
  type: VIEW_WONDER_PANEL,
});

export const viewGameBoard = () => ({
  type: VIEW_GAME_BOARD,
});

export const startAge = () => ({
  type: START_AGE,
});

export const scoreGame = () => ({
  type: SCORE_GAME,
});

export const quit = () => ({
  type: QUIT,
});
