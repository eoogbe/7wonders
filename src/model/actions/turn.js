/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const BUILD_STRUCTURE = 'BUILD_STRUCTURE';

export const BUILD_WONDER = 'BUILD_WONDER';

export const ADD_MONEY = 'ADD_MONEY';

export const SELECT_HAND_CARD = 'SELECT_HAND_CARD';

export const CHOOSE_NEIGHBOR = 'CHOOSE_NEIGHBOR';

export const COPY_GUILD = 'COPY_GUILD';

export const buildStructure = (cardId) => ({
  type: BUILD_STRUCTURE,
  payload: {
    cardId,
  },
});

export const buildFreeStructure = () => ({
  type: BUILD_STRUCTURE,
  payload: {
    isFree: true,
  },
});

export const buildWonder = () => ({
  type: BUILD_WONDER,
});

export const addMoney = () => ({
  type: ADD_MONEY,
});

export const selectHandCard = (cardId, cardIdx) => ({
  type: SELECT_HAND_CARD,
  payload: {
    cardId,
    cardIdx,
  },
});

export const chooseNeighbor = (costType) => ({
  type: CHOOSE_NEIGHBOR,
  payload: {
    costType,
  },
});

export const copyGuild = (cardId) => ({
  type: COPY_GUILD,
  payload: {
    cardId,
  },
});
