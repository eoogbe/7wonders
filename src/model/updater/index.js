/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { concat, decorate, withDefaultState } from 'redux-fp';
import ages from './ages';
import app, { initialAppState } from './app';
import copyAiGuild from './copyAiGuild';
import hands from './hands';
import military from './military';
import neighbors from './neighbors';
import players from './players';
import takeAiTurns from './takeAiTurns';
import trade from './trade';
import wonder from './wonder';

const initialState = {
  ages: [],
  players: {},
  app: initialAppState,
};

export default decorate(
  withDefaultState(initialState),
  concat(
    wonder,
    neighbors,
    takeAiTurns,
    ages,
    hands,
    players,
    military,
    copyAiGuild,
    app,
    trade
  )
);
