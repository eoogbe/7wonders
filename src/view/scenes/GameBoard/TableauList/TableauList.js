/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { compose } from 'ramda';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from 'model/actions';
import { getHasTradeChoice } from 'model/selectors/trade';
import { getTableaus } from 'presentation/tableaus';
import './TableauList.css';
import NeighborChooser from './NeighborChooser';
import Tableau from '../Tableau';

const TableauList = () => {
  const [leftNeighborTableau, playerTableau, rightNeighborTableau] =
    useSelector(getTableaus);
  const hasTradeChoice = useSelector(getHasTradeChoice);

  const dispatch = useDispatch();
  const chooseNeighbor = compose(dispatch, actions.chooseNeighbor);
  const showTableauCallout = compose(dispatch, actions.showTableauCallout);

  const leftTableauStyle = {
    width: '33.3%',
    borderStyle: 'solid',
    borderColor: '#000',
    borderWidth: '0 1px 1px 0',
  };
  const centerTableauStyle = {
    justifyContent: 'center',
    width: '33.3%',
    borderTop: '1px solid #000',
  };
  const rightTableauStyle = {
    justifyContent: 'flex-end',
    width: '33.3%',
    borderStyle: 'solid',
    borderColor: '#000',
    borderWidth: '0 0 1px 1px',
  };

  return (
    <div className="TableauList">
      <Tableau
        {...leftNeighborTableau}
        style={leftTableauStyle}
        showTableauCallout={showTableauCallout}
      />
      {hasTradeChoice && <NeighborChooser chooseNeighbor={chooseNeighbor} />}
      {!hasTradeChoice &&
        <Tableau
          {...playerTableau}
          style={centerTableauStyle}
          showTableauCallout={showTableauCallout}
        />}
      <Tableau
        {...rightNeighborTableau}
        style={rightTableauStyle}
        showTableauCallout={showTableauCallout}
      />
    </div>
  );
};

export default TableauList;
