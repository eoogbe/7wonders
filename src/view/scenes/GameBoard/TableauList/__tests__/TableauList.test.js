/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PLAYERS from '__fixtures__/players';
import { chooseNeighbor, showTableauCallout } from 'model/actions';
import TableauList from '../TableauList';

const MULTIPLE_PAYMENTS = [
  {
    payment: [{ playerId: 'PLAYER2', alternatives: ['Clay'] }],
    leftCost: 2,
    rightCost: 0,
  },
  {
    payment: [{ playerId: 'PLAYER3', alternatives: ['Clay'] }],
    leftCost: 0,
    rightCost: 2,
  },
];

jest.useFakeTimers();

const mockStore = configureStore();

it('renders without crashing', () => {
  const store = mockStore(createState({ structurePayments: [] }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <TableauList />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders with trade choice', () => {
  const store = mockStore(createState({
    structurePayments: MULTIPLE_PAYMENTS,
  }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <TableauList />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('chooses a neighbor', () => {
  const store = mockStore(createState({
    structurePayments: MULTIPLE_PAYMENTS,
  }));
  const tree = mount(
    <Provider store={store}>
      <TableauList />
    </Provider>
  );
  tree.find('.NeighborChooser button').first().simulate('click');
  expect(store.getActions()).toEqual([chooseNeighbor('leftCost')]);
});

it('shows the tableau callout', () => {
  const store = mockStore(createState({ structurePayments: [] }));
  const tree = mount(
    <Provider store={store}>
      <TableauList />
    </Provider>
  );
  tree.find('.TableauCard').first().simulate('mouseEnter');
  expect(store.getActions()).toEqual([
    showTableauCallout('TEMPLE_OF_ARTEMIS_IN_EPHESUS_INITIAL'),
  ]);
});

const createState = ({ structurePayments }) => ({
  app: { currentPlayer: 'PLAYER1' },
  players: {
    ...PLAYERS,
    PLAYER1: {
      ...PLAYERS.PLAYER1,
      structurePayments,
    },
  },
});
