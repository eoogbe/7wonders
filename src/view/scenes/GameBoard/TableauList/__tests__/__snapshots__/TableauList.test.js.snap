// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with trade choice 1`] = `
<div
  className="TableauList"
>
  <div
    className="Tableau"
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": "0 1px 1px 0",
        "width": "33.3%",
      }
    }
  >
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(189,189,189)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Papyrus
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Clay
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(239,83,80)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            1 shield
          </li>
        </ul>
      </li>
    </ul>
  </div>
  <section
    className="NeighborChooser"
  >
    <h2
      className="NeighborChooser-heading"
    >
      Choose a neighbor to trade with:
    </h2>
    <ul
      className="NeighborChooser-buttonPanel"
    >
      <li>
        <button
          className="primary-button"
          onClick={[Function]}
          type="button"
        >
          Left neighbor
        </button>
      </li>
      <li>
        <button
          className="primary-button"
          onClick={[Function]}
          type="button"
        >
          Right neighbor
        </button>
      </li>
    </ul>
  </section>
  <div
    className="Tableau"
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": "0 0 1px 1px",
        "justifyContent": "flex-end",
        "width": "33.3%",
      }
    }
  >
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Ore
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Clay / Ore
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(33,150,243)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            3 points
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(33,150,243)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            2 points
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="TableauList"
>
  <div
    className="Tableau"
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": "0 1px 1px 0",
        "width": "33.3%",
      }
    }
  >
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(189,189,189)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Papyrus
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Clay
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(239,83,80)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            1 shield
          </li>
        </ul>
      </li>
    </ul>
  </div>
  <div
    className="Tableau"
    style={
      Object {
        "borderTop": "1px solid #000",
        "justifyContent": "center",
        "width": "33.3%",
      }
    }
  >
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Stone
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(189,189,189)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Glass
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(239,83,80)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            1 shield
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255,235,59)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            5 coins
          </li>
        </ul>
      </li>
    </ul>
  </div>
  <div
    className="Tableau"
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": "0 0 1px 1px",
        "justifyContent": "flex-end",
        "width": "33.3%",
      }
    }
  >
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Ore
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(141,110,99)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            Clay / Ore
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="Tableau-cardList"
    >
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(33,150,243)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            3 points
          </li>
        </ul>
      </li>
      <li
        className="TableauCard"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(33,150,243)",
            "border": "1px solid rgba(93, 64, 55)",
          }
        }
      >
        <ul
          className="TableauCard-effectList"
        >
          <li
            className="TableauCard-effect"
          >
            2 points
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;
